File: src/platform.ts
import { CrossPlayPreference, Platform } from "./types.js";

const CONSOLE_PLATFORMS: ReadonlySet<Platform> = new Set([
  Platform.XBOX_SERIES,
  Platform.XBOX_ONE,
  Platform.PS5,
  Platform.PS4,
]);

const XBOX_PLATFORMS: ReadonlySet<Platform> = new Set([
  Platform.XBOX_SERIES,
  Platform.XBOX_ONE,
]);

const PLAYSTATION_PLATFORMS: ReadonlySet<Platform> = new Set([
  Platform.PS5,
  Platform.PS4,
]);

export function isPc(platform: Platform): boolean {
  return platform === Platform.PC;
}

export function isConsole(platform: Platform): boolean {
  return CONSOLE_PLATFORMS.has(platform);
}

export function isXbox(platform: Platform): boolean {
  return XBOX_PLATFORMS.has(platform);
}

export function isPlayStation(platform: Platform): boolean {
  return PLAYSTATION_PLATFORMS.has(platform);
}

export function platformFamily(platform: Platform): "pc" | "xbox" | "playstation" {
  if (isPc(platform)) return "pc";
  if (isXbox(platform)) return "xbox";
  return "playstation";
}

/**
 * Returns true when two platforms may appear in the same lobby under the
 * requesting player's cross-play preference.
 */
export function platformsCompatible(
  requesterPlatform: Platform,
  requesterPreference: CrossPlayPreference,
  candidatePlatform: Platform,
): boolean {
  if (requesterPlatform === candidatePlatform) {
    return true;
  }

  switch (requesterPreference) {
    case CrossPlayPreference.ALL:
      return true;

    case CrossPlayPreference.CONSOLE_ONLY:
      // Block PC entirely; allow Xbox <-> PlayStation.
      if (isPc(candidatePlatform)) return false;
      if (isPc(requesterPlatform)) return false;
      return isConsole(candidatePlatform);

    case CrossPlayPreference.SAME_PLATFORM:
      return platformFamily(requesterPlatform) === platformFamily(candidatePlatform);

    default:
      return false;
  }
}

/** Normalize legacy or shorthand platform strings from APIs / telemetry. */
export function parsePlatform(raw: string): Platform | null {
  const normalized = raw.trim().toLowerCase().replace(/[\s-]+/g, "_");

  const map: Record<string, Platform> = {
    pc: Platform.PC,
    win: Platform.PC,
    windows: Platform.PC,
    battlenet: Platform.PC,
    steam: Platform.PC,
    xbox_series: Platform.XBOX_SERIES,
    xbox_series_x: Platform.XBOX_SERIES,
    xbox_series_s: Platform.XBOX_SERIES,
    xbox_scarlett: Platform.XBOX_SERIES,
    xbox_one: Platform.XBOX_ONE,
    xbox: Platform.XBOX_ONE,
    ps5: Platform.PS5,
    playstation_5: Platform.PS5,
    ps4: Platform.PS4,
    playstation_4: Platform.PS4,
    playstation: Platform.PS4,
  };

  return map[normalized] ?? null;
}

File: src/platform.cpp
#include "warzone/platform.hpp"

#include <algorithm>
#include <cctype>
#include <unordered_map>

namespace warzone {

namespace {

std::string normalizeToken(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(),
                 [](unsigned char c) { return static_cast<char>(std::tolower(c)); });

  for (char& c : value) {
    if (c == ' ' || c == '-') {
      c = '_';
    }
  }

  while (!value.empty() && value.front() == '_') {
    value.erase(value.begin());
  }
  while (!value.empty() && value.back() == '_') {
    value.pop_back();
  }

  return value;
}

}  // namespace

bool isPc(Platform platform) { return platform == Platform::Pc; }

bool isConsole(Platform platform) {
  return platform == Platform::XboxSeries || platform == Platform::XboxOne ||
         platform == Platform::Ps5 || platform == Platform::Ps4;
}

bool isXbox(Platform platform) {
  return platform == Platform::XboxSeries || platform == Platform::XboxOne;
}

bool isPlayStation(Platform platform) {
  return platform == Platform::Ps5 || platform == Platform::Ps4;
}

PlatformFamily platformFamily(Platform platform) {
  if (isPc(platform)) {
    return PlatformFamily::Pc;
  }
  if (isXbox(platform)) {
    return PlatformFamily::Xbox;
  }
  return PlatformFamily::PlayStation;
}

bool platformsCompatible(Platform requesterPlatform,
                         CrossPlayPreference requesterPreference,
                         Platform candidatePlatform) {
  if (requesterPlatform == candidatePlatform) {
    return true;
  }

  switch (requesterPreference) {
    case CrossPlayPreference::All:
      return true;
    case CrossPlayPreference::ConsoleOnly:
      if (isPc(candidatePlatform) || isPc(requesterPlatform)) {
        return false;
      }
      return isConsole(candidatePlatform);
    case CrossPlayPreference::SamePlatform:
      return platformFamily(requesterPlatform) == platformFamily(candidatePlatform);
  }

  return false;
}

std::optional<Platform> parsePlatform(const std::string& raw) {
  static const std::unordered_map<std::string, Platform> map = {
      {"pc", Platform::Pc},
      {"win", Platform::Pc},
      {"windows", Platform::Pc},
      {"battlenet", Platform::Pc},
      {"steam", Platform::Pc},
      {"xbox_series", Platform::XboxSeries},
      {"xbox_series_x", Platform::XboxSeries},
      {"xbox_series_s", Platform::XboxSeries},
      {"xbox_scarlett", Platform::XboxSeries},
      {"xbox_one", Platform::XboxOne},
      {"xbox", Platform::XboxOne},
      {"ps5", Platform::Ps5},
      {"playstation_5", Platform::Ps5},
      {"ps4", Platform::Ps4},
      {"playstation_4", Platform::Ps4},
      {"playstation", Platform::Ps4},
  };

  const auto it = map.find(normalizeToken(raw));
  if (it == map.end()) {
    return std::nullopt;
  }
  return it->second;
}

std::optional<CrossPlayPreference> parseCrossPlayPreference(const std::string& raw) {
  static const std::unordered_map<std::string, CrossPlayPreference> map = {
      {"all", CrossPlayPreference::All},
      {"console_only", CrossPlayPreference::ConsoleOnly},
      {"console", CrossPlayPreference::ConsoleOnly},
      {"same_platform", CrossPlayPreference::SamePlatform},
      {"same", CrossPlayPreference::SamePlatform},
  };

  const auto it = map.find(normalizeToken(raw));
  if (it == map.end()) {
    return std::nullopt;
  }
  return it->second;
}

}  // namespace warzone
